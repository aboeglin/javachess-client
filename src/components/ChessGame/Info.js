import React from "react";
import { PropTypes } from "prop-types";

import withChess from "@hocs/withChess";

export const Info = ({ activePlayerColor, playerColor }) => (
  <div>
    <h2>You are {playerColor}</h2>
    <h2>Current turn: {activePlayerColor}</h2>
  </div>
);

Info.propTypes = {
  activePlayerColor: PropTypes.string,
  playerColor: PropTypes.string
};

export default withChess(Info);
