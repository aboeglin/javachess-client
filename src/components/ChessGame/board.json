{
  "squares": [
    { "color": "BLACK", "x": "a", "y": "1" },
    { "color": "WHITE", "x": "b", "y": "1" },
    { "color": "BLACK", "x": "c", "y": "1" },
    { "color": "WHITE", "x": "d", "y": "1" },
    { "color": "BLACK", "x": "e", "y": "1" },
    { "color": "WHITE", "x": "f", "y": "1" },
    { "color": "BLACK", "x": "g", "y": "1" },
    { "color": "WHITE", "x": "h", "y": "1" },

    { "color": "WHITE", "x": "a", "y": "2" },
    { "color": "BLACK", "x": "b", "y": "2" },
    { "color": "WHITE", "x": "c", "y": "2" },
    { "color": "BLACK", "x": "d", "y": "2" },
    { "color": "WHITE", "x": "e", "y": "2" },
    { "color": "BLACK", "x": "f", "y": "2" },
    { "color": "WHITE", "x": "g", "y": "2" },
    { "color": "BLACK", "x": "h", "y": "2" },

    { "color": "BLACK", "x": "a", "y": "3" },
    { "color": "WHITE", "x": "b", "y": "3" },
    { "color": "BLACK", "x": "c", "y": "3" },
    { "color": "WHITE", "x": "d", "y": "3" },
    { "color": "BLACK", "x": "e", "y": "3" },
    { "color": "WHITE", "x": "f", "y": "3" },
    { "color": "BLACK", "x": "g", "y": "3" },
    { "color": "WHITE", "x": "h", "y": "3" },

    { "color": "WHITE", "x": "a", "y": "4" },
    { "color": "BLACK", "x": "b", "y": "4" },
    { "color": "WHITE", "x": "c", "y": "4" },
    { "color": "BLACK", "x": "d", "y": "4" },
    { "color": "WHITE", "x": "e", "y": "4" },
    { "color": "BLACK", "x": "f", "y": "4" },
    { "color": "WHITE", "x": "g", "y": "4" },
    { "color": "BLACK", "x": "h", "y": "4" },

    { "color": "BLACK", "x": "a", "y": "5" },
    { "color": "WHITE", "x": "b", "y": "5" },
    { "color": "BLACK", "x": "c", "y": "5" },
    { "color": "WHITE", "x": "d", "y": "5" },
    { "color": "BLACK", "x": "e", "y": "5" },
    { "color": "WHITE", "x": "f", "y": "5" },
    { "color": "BLACK", "x": "g", "y": "5" },
    { "color": "WHITE", "x": "h", "y": "5" },

    { "color": "WHITE", "x": "a", "y": "6" },
    { "color": "BLACK", "x": "b", "y": "6" },
    { "color": "WHITE", "x": "c", "y": "6" },
    { "color": "BLACK", "x": "d", "y": "6" },
    { "color": "WHITE", "x": "e", "y": "6" },
    { "color": "BLACK", "x": "f", "y": "6" },
    { "color": "WHITE", "x": "g", "y": "6" },
    { "color": "BLACK", "x": "h", "y": "6" },

    { "color": "BLACK", "x": "a", "y": "7" },
    { "color": "WHITE", "x": "b", "y": "7" },
    { "color": "BLACK", "x": "c", "y": "7" },
    { "color": "WHITE", "x": "d", "y": "7" },
    { "color": "BLACK", "x": "e", "y": "7" },
    { "color": "WHITE", "x": "f", "y": "7" },
    { "color": "BLACK", "x": "g", "y": "7" },
    { "color": "WHITE", "x": "h", "y": "7" },

    { "color": "WHITE", "x": "a", "y": "8" },
    { "color": "BLACK", "x": "b", "y": "8" },
    { "color": "WHITE", "x": "c", "y": "8" },
    { "color": "BLACK", "x": "d", "y": "8" },
    { "color": "WHITE", "x": "e", "y": "8" },
    { "color": "BLACK", "x": "f", "y": "8" },
    { "color": "WHITE", "x": "g", "y": "8" },
    { "color": "BLACK", "x": "h", "y": "8" }
  ]
}
