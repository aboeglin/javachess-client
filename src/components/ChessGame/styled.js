import styled from "@emotion/styled";

const SQUARE_WIDTH = "12.5%";
const SQUARE_HEIGHT = "12.5%";

const WHITE = "#f2efeb";
const BLACK = "#331c13";

export const Container = styled.div`
  width: 640px;
  height: 640px;
  background-color: blue;
  position: relative;
  margin: auto;

  h2 {
    margin: 0;
  }
`;

export const Square = styled.div`
  position: relative;
  display: inline-block;
  width: ${SQUARE_WIDTH};
  height: ${SQUARE_HEIGHT};
  margin: 0;

  background-color: ${props => (props.color === "WHITE" ? WHITE : BLACK)};
  box-shadow: inset 0px 0px 26px 5px rgba(0, 0, 0, 0.2);
`;

export const SquareLabel = styled.div`
  color: ${props => (props.color === "WHITE" ? BLACK : WHITE)};
  opacity: 0.3;
  margin-left: 5px;
  margin-top: 5px;
`;

export const PieceContainer = styled.div`
  height: 100%;

  img {
    height: ${props => {
      switch (props.type) {
        case "QUEEN":
        case "KING":
        case "BISHOP":
        case "ROOK":
        case "KNIGHT":
          return "65px";
        default:
          return "55px";
      }
    }};
    position: absolute;
    display: block;
    bottom: 5px;
    left: 50%;
    transform: ${props => {
      console.log(props);
      switch (props.x) {
        case "a":
        case "b":
        case "c":
        case "d":
          return "translateX(-50%) scaleX(-1)";
        default:
          return "translateX(-50%)";
      }
    }};
  }
`;
