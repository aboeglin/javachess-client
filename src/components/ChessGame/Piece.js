import React from "react";

import { curry, pipe } from "ramda";

import { PieceContainer } from "./styled";

import BLACK_PAWN_IMAGE_URL from "./images/pawn-black.png";
import BLACK_BISHOP_IMAGE_URL from "./images/bishop-black.png";
import BLACK_KNIGHT_IMAGE_URL from "./images/knight-black.png";
import BLACK_ROOK_IMAGE_URL from "./images/rook-black.png";
import BLACK_QUEEN_IMAGE_URL from "./images/queen-black.png";
import BLACK_KING_IMAGE_URL from "./images/king-black.png";
import WHITE_PAWN_IMAGE_URL from "./images/pawn-white.png";
import WHITE_BISHOP_IMAGE_URL from "./images/bishop-white.png";
import WHITE_KNIGHT_IMAGE_URL from "./images/knight-white.png";
import WHITE_ROOK_IMAGE_URL from "./images/rook-white.png";
import WHITE_QUEEN_IMAGE_URL from "./images/queen-white.png";
import WHITE_KING_IMAGE_URL from "./images/king-white.png";
import withChess from "@hocs/withChess";

const getImageUrl = (type, color) =>
  ({
    BLACK_PAWN: BLACK_PAWN_IMAGE_URL,
    BLACK_BISHOP: BLACK_BISHOP_IMAGE_URL,
    BLACK_KNIGHT: BLACK_KNIGHT_IMAGE_URL,
    BLACK_ROOK: BLACK_ROOK_IMAGE_URL,
    BLACK_QUEEN: BLACK_QUEEN_IMAGE_URL,
    BLACK_KING: BLACK_KING_IMAGE_URL,
    WHITE_PAWN: WHITE_PAWN_IMAGE_URL,
    WHITE_BISHOP: WHITE_BISHOP_IMAGE_URL,
    WHITE_KNIGHT: WHITE_KNIGHT_IMAGE_URL,
    WHITE_ROOK: WHITE_ROOK_IMAGE_URL,
    WHITE_QUEEN: WHITE_QUEEN_IMAGE_URL,
    WHITE_KING: WHITE_KING_IMAGE_URL
  }[`${color}_${type}`]);

const handlePieceClicked = curry(
  (piece, playerColor, activePlayerColor, pieceSelected) => () => {
    if (piece.color === playerColor && activePlayerColor === playerColor) {
      pieceSelected(piece);
    }
  }
);

export const Piece = ({
  piece,
  pieceSelected,
  playerColor,
  activePlayerColor
}) =>
  pipe(
    ({ type, color }) => getImageUrl(type, color),
    url => (
      <PieceContainer
        type={piece.type}
        x={piece.x}
        onClick={handlePieceClicked(
          piece,
          playerColor,
          activePlayerColor,
          pieceSelected
        )}
      >
        <img src={url} />
      </PieceContainer>
    )
  )(piece);

export default withChess(Piece);
