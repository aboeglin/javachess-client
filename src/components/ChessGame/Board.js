import React from "react";
import { PropTypes } from "prop-types";
import {
  always,
  curry,
  pipe,
  map,
  reverse,
  flatten,
  find,
  splitEvery,
  ifElse,
  identity
} from "ramda";
import { Square, SquareLabel } from "./styled";
import Piece from "./Piece";
import BOARD from "./board.json";
import withChess from "../../hoc/withChess";

// Move to separate file
const SquareContainer = ({ square, piece }) => (
  <Square
    color={square.color}
    x={square.x}
    y={square.y}
    key={`${square.x}${square.y}`}
  >
    <SquareLabel color={square.color}>
      {square.x}
      {square.y}
    </SquareLabel>
    {piece}
  </Square>
);

SquareContainer.propTypes = {
  square: PropTypes.object,
  piece: PropTypes.object
};

const orderSquaresForDisplay = pipe(
  reverse,
  splitEvery(8),
  map(reverse),
  flatten
);

const addPiece = curry((pieces, square) => ({
  square,
  piece: pieces.length ? getPieceAtSquare(square, pieces) : null
}));

const Board = ({ squares, pieces }) =>
  pipe(
    map(addPiece(pieces)),
    map(SquareContainer),
    orderSquaresForDisplay
  )(squares);

const getPieceAtSquare = curry((square, pieces) =>
  pipe(
    find(({ x, y }) => x === square.x && y === square.y),
    ifElse(identity, piece => <Piece {...piece} />, always(null))
  )(pieces)
);

Board.propTypes = {
  squares: PropTypes.array,
  pieces: PropTypes.array
};

Board.defaultProps = {
  squares: BOARD.squares,
  pieces: []
};

export default withChess(Board);
