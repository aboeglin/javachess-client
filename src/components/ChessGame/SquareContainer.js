import React from "react";
import { PropTypes } from "prop-types";
import { curry, find, pipe, propOr } from "ramda";

import Piece from "./Piece";
import { Square, SquareLabel } from "./styled";
import withChess from "../../hoc/withChess";

const getPieceAtSquare = curry((pieces, square) =>
  find(({ x, y }) => x === square.x && y === square.y)(pieces)
);

const isSelected = curry(
  (selection, square) =>
    square.x === propOr(false, "x", selection) &&
    square.y === propOr(false, "y", selection)
);

const getSelectionColor = selection => (selection ? selection.color : false);

export const SquareContainer = ({ square, pieces, pieceMoved, selection }) =>
  pipe(getPieceAtSquare(pieces), piece => (
    <Square
      color={square.color}
      x={square.x}
      y={square.y}
      key={`${square.x}${square.y}`}
      highlighted={square.highlighted}
      selected={isSelected(selection, square)}
      selectionColor={getSelectionColor(selection)}
      onClick={() => {
        if (
          selection &&
          (selection.x !== square.x || selection.y !== square.y)
        ) {
          if (!piece || piece.color !== selection.color) {
            pieceMoved(selection.x, selection.y, square.x, square.y);
          }
        }
      }}
    >
      <SquareLabel color={square.color}>
        {square.x}
        {square.y}
      </SquareLabel>
      {piece && <Piece piece={piece} />}
    </Square>
  ))(square);

SquareContainer.propTypes = {
  square: PropTypes.object,
  pieces: PropTypes.array
};

export default withChess(SquareContainer);
