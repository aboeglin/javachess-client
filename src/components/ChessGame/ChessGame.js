import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { find, pathOr, propEq } from "ramda";
import { fork } from "fluture";

import { Container } from "./styled";

import Board from "@components/Board";
import Info from "./Info";

import { fetchGames, joinGame } from "@utils/upstreams";
import withChess from "@hocs/withChess";

export const ChessGame = ({ startGame, userName }) => {
  const id = parseInt(useParams().id, 10);
  const [games, setGames] = useState([]);
  const [joined, setJoined] = useState(false);

  useEffect(() => {
    fork(console.error)(setGames)(fetchGames());
  }, []);

  const currentGame = find(propEq("id", id))(games);

  useEffect(() => {
    const player1Id = pathOr(null, ["player1", "id"])(currentGame);
		const player2Id = pathOr(null, ["player2", "id"])(currentGame);
		
    const shouldJoin =
      userName !== player1Id && userName !== player2Id && currentGame;
    if (shouldJoin) {
      fork(console.error)(() => setJoined(true))(joinGame(userName, id));
    } else {
      setJoined(true);
    }
  }, [games]);

  useEffect(() => {
    if (joined) {
			startGame(id);
    }
  }, [joined]);

  if (!currentGame) {
    return <div>Getting game data ...</div>;
  } else if (!joined) {
    return <div>Joining game ...</div>;
  } else {
    return (
      <>
        <Info />
        <Container>
          <Board />
        </Container>
      </>
    );
  }
};

export default withChess(ChessGame);
