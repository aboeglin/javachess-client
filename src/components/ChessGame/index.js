import ChessGame from "./ChessGame";

export default ChessGame;
