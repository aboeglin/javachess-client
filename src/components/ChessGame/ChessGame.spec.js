import ChessGame from "./ChessGame";
import { standardComponentTest } from "../../utils/testing";

standardComponentTest(ChessGame, {});
