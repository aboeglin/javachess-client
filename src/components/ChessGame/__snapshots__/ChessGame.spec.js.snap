// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <A Component /> 1`] = `
<div>
  Getting game data ...
</div>
`;
